cmake/multi_image.cmake

```diff
@@ -0,0 +1,164 @@
+  set(${UPNAME}_PROJECT_BINARY_DIR ${CMAKE_BINARY_DIR}/${name}/zephyr)
+
+  get_cmake_property(VARIABLES_CACHED CACHE_VARIABLES)
+  get_cmake_property(VARIABLES              VARIABLES)
+
+  unset(image_cmake_args)
+  foreach(var_name
+      ${VARIABLES}
+      ${VARIABLES_CACHED}
+      )
+    if("${var_name}" MATCHES "^${name}_(.+)") # TODO: How expensive?
+      list(APPEND image_cmake_args
+        -D${CMAKE_MATCH_1}=${${var_name}}
+        )
+    endif()
+  endforeach()
+
+  file(MAKE_DIRECTORY ${${UPNAME}_CMAKE_BINARY_DIR})
+  message("=== child image ${name} begin ===")
+  image_board_selection(${BOARD} IMAGE_BOARD)
+  image_menuconfig(IMAGE_KCONFIG_TARGET IMAGE_COMMAND_TARGET)
+  execute_process(
+    COMMAND ${CMAKE_COMMAND}
+    # Add other toolchain here,
+    -G${CMAKE_GENERATOR}
+    -DBOARD=${IMAGE_BOARD}
+    ${IMAGE_KCONFIG_TARGET}
+    ${IMAGE_COMMAND_TARGET}
+    -DIS_CHILD_IMAGE=True
+    ${sourcedir}
+    ${image_cmake_args}
+    WORKING_DIRECTORY ${${UPNAME}_CMAKE_BINARY_DIR}
+    RESULT_VARIABLE ret
+    )
+  set(${UPNAME}_DOTCONFIG ${${UPNAME}_PROJECT_BINARY_DIR}/.config)
+  set_property(DIRECTORY APPEND PROPERTY CMAKE_CONFIGURE_DEPENDS ${${UPNAME}_DOTCONFIG})
+
+  if(NOT ${ret} EQUAL "0")
+    message(FATAL_ERROR "CMake generation for ${name} failed, aborting. Command: ${ret}")
+  endif()
+
+  message("=== child image ${name} end ===")
+
+  # The variables inhereted from PARENT_SCOPE might be overwritten when additional Kconfig is loaded locally.
+  # Therefore, safe guard the variables from parent scope that are needed.
+  import_kconfig(CONFIG_ ${${UPNAME}_PROJECT_BINARY_DIR}/.config)
+
+  # Clear it, just in case it was defined in parent scope
+  set(spm_veneers_lib)
+  if(CONFIG_${UPNAME}_CONTAINS_SECURE_ENTRIES)
+    set(spm_veneers_lib ${${UPNAME}_CMAKE_BINARY_DIR}/${CONFIG_ARM_ENTRY_VENEERS_LIB_NAME})
+  endif()
+
+  define_property(GLOBAL PROPERTY ${name}_KERNEL_NAME
+                 BRIEF_DOCS "Name (KERNEL_NAME) for the SPM image."
+                 FULL_DOCS "To be used to access e.g. this image's hex file."
+                )
+  set_property(GLOBAL PROPERTY ${name}_KERNEL_NAME ${CONFIG_KERNEL_BIN_NAME})
+
+  include(ExternalProject)
+  ExternalProject_Add(${UPNAME}_subimage
+      SOURCE_DIR ${sourcedir}
+      BINARY_DIR ${${UPNAME}_CMAKE_BINARY_DIR}
+      BUILD_BYPRODUCTS
+       # TODO: Remove this coupling
+      ${spm_veneers_lib}
+      ${${UPNAME}_PROJECT_BINARY_DIR}/${CONFIG_KERNEL_BIN_NAME}.hex
+      # The CMake configure has been executed above due to cross
+      # dependencies between the multiple CMake configure stages.
+      CONFIGURE_COMMAND ""
+      BUILD_COMMAND ${CMAKE_MAKE_PROGRAM}
+      INSTALL_COMMAND ""
+      BUILD_ALWAYS True
+    )
+
+    # ToDo: Consider to use a dedicated target and populate that target directly
+    # with the files as names and deps.
+    # That would allow to cleanup logic found in partition_manager.cmake
```

Comment:
I completely agree, that using a custom target and populate it, instead of the global conf used currently is much nicer.
It also allows for moving CMake code around, whereas the current use of global conf is way to fragile for execution order changes.

Question is: should that be done separately, in order to keep scope of this PR limited.

Reply:
I don't know, I haven't reviewed this code yet.